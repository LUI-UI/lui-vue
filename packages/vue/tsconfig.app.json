{
  "extends": "@vue/tsconfig/tsconfig.web.json",
  "compilerOptions": {
    "composite": true,
    "baseUrl": ".",
    "paths": {
      "@*": ["./src/*"]
    }
  },
  "include": ["env.d.ts", "src/**/*", "src/**/*.vue", "tailwind.config.cjs"],
  "exclude": ["src/**/__tests__/*"]
}
