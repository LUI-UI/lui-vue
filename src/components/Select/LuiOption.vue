<script lang="ts">
import { inject, computed, watch, nextTick, useSlots } from 'vue'
import type { PropType } from 'vue'
import { ContextKey } from './symbols'
import { useId } from '../../utils/useId'
import type { Rounded, Size } from '@/globals/types'
import type { TwClassInterface } from '@/globals/interfaces'

export default {
  name: 'LuiOption',
  inheritAttrs: false
}
</script>
<script setup lang="ts">
const props = defineProps({
  selected: {
    type: Boolean,
    default: false
  },
  disabled: {
    type: Boolean,
    default: false
  },
  size: {
    type: String as PropType<Size>,
    default: 'md'
  },
  rounded: {
    type: [Boolean, String] as PropType<Rounded>,
    default: false
  },
  value: {
    type: [String, Number],
    default: ''
  },
  text: {
    type: String,
    default: ''
  }
})

// const { selectedOption, updateSelectedOption, focusButton } =
//   inject(ContextKey);
const context = inject(ContextKey)
const slots = useSlots()
const optionId = `lui-option-${useId()}`

nextTick(() => {
  // if v-model does not used so we set the initial selectedValue
  if (context?.selectedOption.value === undefined && props.selected && !props.disabled) {
    context?.updateSelectedOption({
      value: props.value,
      text: props.text,
      selected: props.selected
    })
  }
})

watch(
  () => props.selected,
  (value) => {
    context?.updateSelectedOption({
      value: props.value,
      text: props.text,
      selected: value
    })
  }
)

function handleOptionClick(e: any) {
  e.preventDefault()
  context?.updateSelectedOption({
    value: props.value,
    text: props.text,
    selected: props.selected
  })
  nextTick(() => context?.focusButton())
}

const isSelected = computed(() => {
  if (context?.selectedOption.value === undefined) return props.selected
  return typeof context?.selectedOption.value === 'string'
    ? props.text === context?.selectedOption.value
    : props.text === context?.selectedOption.value.text
})

const optionClasses = computed(() => {
  //hover:bg-${props.color}-600/20
  const classes: TwClassInterface = {
    backgroundColor: props.disabled
      ? ''
      : isSelected.value === true
      ? 'bg-primary-500 hover:text-primary-600 focus:text-primary-600'
      : 'hover:bg-primary-500/20 focus:bg-primary-500/20',
    padding: {
      'px-1.5 py-1': props.size === 'xs',
      'px-2 py-1.5': props.size === 'sm',
      'px-2 py-2.5': props.size === 'md',
      'p-2.5': props.size === 'lg',
      'px-3 py-3.5': props.size === 'xl'
    },
    textColor: props.disabled
      ? 'text-secondary-300 dark:text-secondary-700'
      : isSelected.value === true
      ? 'text-white'
      : 'text-secondary-600 dark:text-secondary-300 hover:text-primary-500 focus:text-primary-500',
    outlineWidth: 'outline-none',
    // margin: props.size === "xs" || props.size === "sm" ? "mb-1.5" : "mb-2",
    borderRadius: {
      'rounded-md': props.rounded === true,
      'rounded-2xl': props.rounded === 'full'
    },
    display: slots.prepend || slots.append ? 'flex' : '',
    alignItems: slots.prepend || slots.append ? 'items-center' : '',
    cursor: props.disabled ? 'cursor-not-allowed' : '',
    pointerEvents: props.disabled ? 'pointer-events-none' : ''
  }
  return Object.values({ ...classes })
})
</script>
<template>
  <li
    role="option"
    tabindex="-1"
    v-bind="$attrs"
    :id="optionId"
    :aria-selected="isSelected"
    :class="optionClasses"
    @click="handleOptionClick"
  >
    <span v-if="$slots.prepend"> <slot name="prepend" /> </span>
    <span :class="$slots.prepend ? 'ml-1' : ''"> {{ text }} </span>
    <span v-if="$slots.append" class="ml-auto">
      <slot name="append" />
    </span>
    <!-- <span><slot /></span> -->
  </li>
</template>
