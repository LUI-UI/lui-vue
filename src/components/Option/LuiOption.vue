<script lang="ts">
import { computed, inject, nextTick, ref, useSlots, watch } from 'vue'
import type { PropType } from 'vue'
import { ContextKey } from '../Select/symbols'
import { useId } from '../../utils/useId'
import type { Rounded, Size } from '@/globals/types'
import type { TwClassInterface } from '@/globals/interfaces'

export default {
  name: 'LuiOption',
  inheritAttrs: false,
}
</script>

<script setup lang="ts">
const props = defineProps({
  selected: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  size: {
    type: String as PropType<Size>,
    default: 'md',
  },
  rounded: {
    type: [Boolean, String] as PropType<Rounded>,
    default: false,
  },
  value: {
    type: [String, Number],
    default: '',
  },
  text: {
    type: String,
    default: '',
  },
})

// const { selectedOption, updateSelectedOption, focusButton } =
//   inject(ContextKey);
const context = inject(ContextKey)
const slots = useSlots()
const optionId = `lui-option-${useId()}`
const optionRef = ref(null)
const isKeyboardHovered = computed(() => optionId === context?.currentId.value)
nextTick(() => {
  // if v-model does not used so we set the initial selectedValue
  if (context?.selectedOption.value === undefined && props.selected && !props.disabled) {
    context?.updateSelectedOption({
      value: props.value,
      text: props.text,
      selected: props.selected,
    })
  }
})

watch(
  () => props.selected,
  (value) => {
    context?.updateSelectedOption({
      value: props.value,
      text: props.text,
      selected: value,
    })
  },
)

function handleOptionClick(e: any) {
  e.preventDefault()
  const isSelectMultiple = Array.isArray(context?.selectedOption.value)
  if (isSelectMultiple)
    e.stopPropagation()
  context?.updateSelectedOption({
    value: props.value,
    text: props.text,
    selected: props.selected,
  })
  nextTick(() => context?.focusButton())
}
function anySelectedOption() {
  if (Array.isArray(context?.selectedOption.value))
    return context?.selectedOption.value.length !== 0
  return context?.selectedOption.value.text !== ''
}
function isSelectedsIncludesOption() {
  if (!Array.isArray(context?.selectedOption.value))
    return false
  return context?.selectedOption.value.includes(props.text) || context?.selectedOption.value.includes(props.value as string)
}
const isSelected = computed(() => {
  if (!anySelectedOption())
    return props.selected
  return Array.isArray(context?.selectedOption.value)
    ? isSelectedsIncludesOption()
    : context?.selectedOption.value.text === props.text
})
const optionClasses = computed(() => {
  // hover:bg-${props.color}-600/20
  const classes: TwClassInterface = {
    // backgroundColor: props.disabled
    //   ? ''
    //   : isSelected.value
    //     ? isKeyboardHovered.value
    //       ? 'bg-danger-500'
    //       : 'hover:bg-primary-500/20 focus-visible:bg-primary-500/20'
    //     : '',
    // default: '', hover
    // selected: '', hover
    // keybpardHover: selected ? '' : ''

    backgroundColor: props.disabled
      ? ''
      : {
          'hover:bg-primary-500/20': !isSelected.value,
          'bg-primary-500 hover:bg-primary-500/80 focus-visible:bg-primary-500/80': isSelected.value && !isKeyboardHovered.value,
          'bg-primary-500/20': isKeyboardHovered.value && !isSelected.value,
          'bg-primary-500/80': isKeyboardHovered.value && isSelected.value,
        },
    padding: {
      'py-1 px-1.5': props.size === 'xs',
      'py-1.5 px-2': props.size === 'sm',
      'py-2 px-2.5': props.size === 'md',
      'py-2.5 px-2.5': props.size === 'lg',
      'py-3.5 px-3': props.size === 'xl',
    },
    fontSize: {
      'text-xs': props.size === 'xs',
      'text-sm': props.size === 'sm',
      'text-base': props.size === 'md',
      'text-lg': props.size === 'lg',
      'text-xl': props.size === 'xl',
    },
    textColor: props.disabled
      ? 'text-secondary-300 dark:text-secondary-700'
      : {
          'text-secondary-600 dark:text-secondary-300 hover:text-primary-500 focus-visible:text-primary-500': !isKeyboardHovered.value && !isSelected.value, // default state
          'text-white': isSelected.value, // selected
          'text-primary-500': isKeyboardHovered.value && !isSelected.value, // keyboard-hovered && not-selected
        },
    outlineWidth: 'outline-none',
    borderRadius: {
      'rounded-md': props.rounded === true,
      'rounded-2xl': props.rounded === 'full',
    },
    display: slots.prepend || slots.append ? 'flex' : '',
    alignItems: slots.prepend || slots.append ? 'items-center' : '',
    cursor: props.disabled ? 'cursor-not-allowed' : 'cursor-pointer',
    pointerEvents: props.disabled ? 'pointer-events-none' : '',
  }
  return Object.values({ ...classes })
})
defineExpose({
  el: optionRef,
})
</script>

<template>
  <li
    v-bind="$attrs"
    :id="optionId"
    ref="optionRef"
    role="option"
    tabindex="-1"
    :aria-selected="isSelected"
    class="lui-option"
    :class="optionClasses"
    @click="handleOptionClick"
  >
    <span v-if="$slots.prepend"> <slot name="prepend" /> </span>
    <span :class="$slots.prepend ? 'ml-1' : ''"> {{ text }} </span>
    <span v-if="$slots.append" class="ml-auto">
      <slot name="append" />
    </span>
  </li>
</template>
