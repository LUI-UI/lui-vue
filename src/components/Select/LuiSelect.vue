<script lang="ts">
export default {
  name: 'LuiSelect',
  inheritAttrs: false
}
</script>
<script setup lang="ts">
import { ref, nextTick, provide, useSlots, useAttrs, watch, reactive, computed } from 'vue'
import type { PropType, Ref } from 'vue'
import type { OptionsType, ModelValue, ModelValueObject, ListboxStateType } from './select-types'
import { ContextKey } from './symbols'
import { useId } from '../../utils/useId'
// import { useOutsideClick, useFindProperPosition } from "./composables/index";
import { useOutsideClick } from '../../composables/useOutsideClick'
import { useFindProperPosition } from '../../composables/useFindProperPosition'
import type { TwClassInterface } from '@/globals/interfaces'
import type { Rounded, Block, Size, State, StateIcon, Description } from '@/globals/types'
import LuiOption from './LuiOption.vue'
import LuiInput from '../Input/LuiInput.vue'
const props = defineProps({
  rounded: {
    type: [Boolean, String] as PropType<Rounded>,
    default: false
  },
  block: {
    type: Boolean as PropType<Block>,
    default: false
  },
  state: {
    type: [String, Boolean, null] as PropType<State>,
    default: null
  },
  stateIcon: {
    type: [Boolean] as PropType<StateIcon>,
    default: null
  },
  options: {
    type: Array as PropType<OptionsType>,
    default: () => []
  },
  placeholder: {
    type: String as PropType<string>,
    default: ''
  },
  size: {
    type: String as PropType<Size>,
    default: 'md'
  },
  description: {
    type: [String, null] as PropType<Description>,
    default: null
  },
  modelValue: {
    type: [Object, String, undefined] as PropType<ModelValue>,
    default: undefined
  }
})
const slots = useSlots()
const attrs = useAttrs()
// const selectRef: Ref<InstanceType<typeof LuiInput> | null> = ref(null);
const selectRef = ref<InstanceType<typeof LuiInput>>()
const optionsRef = ref<HTMLUListElement>()
const selectWrapperRef = ref<HTMLDivElement>()
const optionsActive: Ref<boolean> = ref(false)
const selectedOption: Ref<any> = ref(undefined)
// const selectedOption: Ref<string | ModelValueObject | undefined> =
//   ref(undefined);
const listboxState: ListboxStateType = reactive({
  items: [],
  currentIndex: 0,
  currentId: ''
})

const selectId = `lui-listbox-button-${useId()}`
const optionsId = `lui-listbox-wrapper-${useId()}`
const validSlotTypes = ['LuiOption']
const errorMessages = {
  type: {
    modelValue: 'Wrong type for modelValue, typeof of modelValue should be string'
  },
  missing: {
    options: `Options missing: should use options prop or LuiOption component as slot`
  }
}

const emit = defineEmits(['update:modelValue', 'change'])

const { properPosition } = useFindProperPosition(selectWrapperRef)
useOutsideClick(selectWrapperRef, () => closeListBox())

// const vClickOutSide = {
//   mounted: function (el: any, binding: any, vnode) {
//     el.clickOutsideEvent = function (event) {
//       if (!(el == event.target || el.contains(event.target))) {
//         binding.value(event, el);
//       }
//     };
//     document.addEventListener("click", el.clickOutsideEvent);
//   },
//   unmounted: function (el) {
//     document.removeEventListener("click", el.clickOutsideEvent);
//   },
// };

nextTick(() => {
  setInitialSelectedOption()
  setState()
})

provide(ContextKey, {
  selectedOption,
  updateSelectedOption,
  focusButton
})

watch(
  () => props.modelValue,
  (value) => {
    const rawValue = typeof value !== 'string' ? value?.text : value
    if (rawValue !== selectedOption.value?.text) {
      updateSelectedOption(value)
    }
    // updateSelectedOption(value);
  }
)

function focusAvailableElement(
  el: HTMLElement | undefined,
  oparation: (i: number) => number,
  initial: number | null = null
) {
  const isTargetExist = (index: number) => index >= 0 && index <= listboxState.items.length - 1
  const isTargetFocusable = (targetIndex: number) => {
    const target = listboxState.items[targetIndex]
    if (
      typeof target !== 'string' &&
      (target?.disabled === undefined || target?.disabled === false)
    )
      return true
    return false
  }
  // listboxState.items[targetIndex]?.disabled === undefined ||
  //   listboxState.items[targetIndex]?.disabled === false;

  let targetIndex = listboxState.currentIndex
  if (initial !== null) {
    targetIndex = initial
  } else {
    targetIndex = oparation(targetIndex)
  }
  if (!isTargetExist(targetIndex)) return

  while (!isTargetFocusable(targetIndex)) {
    targetIndex = oparation(targetIndex)
    if (!isTargetExist(targetIndex)) return
  }

  listboxState.currentIndex = targetIndex
  const currentEl = el?.children[listboxState.currentIndex]
  listboxState.currentId = currentEl?.id

  nextTick(() => (currentEl as HTMLElement)?.focus({ preventScroll: true }))
}

function updateSelectedOption(option: ModelValue) {
  selectedOption.value = option
  let emitedVal = option
  if (typeof option !== 'string') {
    emitedVal = option?.text
  }
  emit('update:modelValue', emitedVal)
  emit('change', emitedVal)
}

function focusButton() {
  // selectRef.value?.focus({ preventScroll: true });
  const select = selectRef.value as any
  select?.focus()
}

function closeListBox() {
  optionsActive.value = false
}

function toggleOptions() {
  if (attrs.disabled !== undefined && attrs.disabled === true) return
  // event.preventDefault();
  optionsActive.value = !optionsActive.value
  // if (attrs.disabled !== undefined && attrs.disabled === true) return;
}

function setState() {
  const slotsOptions =
    slots.default &&
    slots
      .default()
      .map((slot: any) =>
        slot.type.toString() === 'Symbol(Fragment)'
          ? slot.children.map((child: any) => child.props)
          : slot.props
      )
      .flat()
  // validSlotTypes.includes(child.type.name)
  const allOptions = [...props.options].concat(slotsOptions || [])
  listboxState.items = allOptions
}

function setInitialSelectedOption() {
  // when options not provided also there is no slot we need the throw an error like you should use options prop or option component as slot.
  // const isModelValueInvalid =
  //   props.modelValue !== undefined &&
  //   typeof props.modelValue !== "string" &&
  //   (props.modelValue?.text === undefined ||
  //     props.modelValue?.value === undefined);
  const isModelValueInvalid =
    props.modelValue !== undefined &&
    typeof props.modelValue !== 'string' &&
    typeof props.modelValue !== 'number'

  const optionsExist = props.options.length > 0

  const anyOptionSelected = () =>
    props.options.some((o: ModelValueObject | string) => typeof o !== 'string' && o.selected)

  function setPlaceholderOrValue(value: any) {
    if (props.placeholder === '') {
      updateSelectedOption(value)
    } else {
      // updateSelectedOption(props.placeholder);
    }
  }
  const anySlotSelected = () =>
    slots.default &&
    slots
      .default()
      .some((slot: any) =>
        slot.type.toString() == 'Symbol(Fragment)'
          ? slot.children.some(
              (child: any) => child.props.selected !== undefined && child.props.selected === true
            )
          : slot.props && slot.props.selected && slot.props.selected === true
      )

  const isDefaultSlotValid = () =>
    slots.default &&
    slots
      .default()
      .some((slot: any) =>
        slot.type.toString() == 'Symbol(Fragment)'
          ? slot.children.some(
              (child: any) =>
                child.type.name !== undefined && validSlotTypes.includes(child.type.name)
            )
          : slot.type.name !== undefined && validSlotTypes.includes(slot.type.name)
      )

  if (isModelValueInvalid) {
    throw new Error(errorMessages.type.modelValue)
  }
  if (props.modelValue !== undefined) {
    // should we handle the case if modelValue does not match any option then if placeholder exist set placeholder than throw error?
    updateSelectedOption(props.modelValue)
    return
  }
  // after this line modelValue is undefined
  if (optionsExist && !anyOptionSelected()) {
    setPlaceholderOrValue(props.options[0])
    return
  }

  if (!optionsExist && !isDefaultSlotValid()) {
    throw new Error(
      `Options missing: should use options prop or one of the valid slots: ${validSlotTypes}`
    )
  }
  if (!optionsExist && !anySlotSelected()) {
    const firstSlot: any = slots.default && slots.default()[0]
    let propsOfFirstSlot
    if (slots.default && firstSlot?.type.toString() === 'Symbol(Fragment)') {
      propsOfFirstSlot = firstSlot?.children[0]?.props
    } else {
      propsOfFirstSlot = firstSlot?.props
    }
    setPlaceholderOrValue(propsOfFirstSlot)
  }
}

function buttonKeydown(event: KeyboardEvent) {
  switch (event.code) {
    case 'ArrowDown':
    case 'ArrowUp':
    case 'Enter':
    case 'Space':
      {
        event.preventDefault()
        if (!optionsActive.value) {
          toggleOptions()
        }
        // nextTick(() => Focus(optionsRef.value, "selected"));
        // if selected item exist
        // check is selected item disabled (it is possible for some case)
        // if not disabled focus the selected item
        // if disabled focus next element
        // if not exist
        // focus first item
        let selectedIndex = listboxState.items.findIndex((item: any) =>
          typeof item === 'string'
            ? item === selectedOption.value
            : item?.text === selectedOption.value?.text
        )
        if (selectedIndex === -1) {
          focusAvailableElement(optionsRef.value, (i) => i + 1, 0)
        } else {
          focusAvailableElement(optionsRef.value, (i) => i + 1, selectedIndex)
        }
      }

      break
    default:
    // code block
  }
}
// function handleKeyUp(event: KeyboardEvent) {
//   switch (event.key) {
//     case Keys.Space:
//       // Required for firefox, event.preventDefault() in handleKeyDown for
//       // the Space key doesn't cancel the handleKeyUp, which in turn
//       // triggers a *click*.
//       event.preventDefault();
//       break;
//   }
// }

function optionsKeydown(event: KeyboardEvent) {
  switch (event.code) {
    case 'ArrowDown':
      event.preventDefault()
      focusAvailableElement(optionsRef.value, (i) => i + 1)
      break
    case 'ArrowUp':
      event.preventDefault()
      focusAvailableElement(optionsRef.value, (i) => i - 1)
      break
    case 'Enter':
      event.preventDefault()
      event.stopPropagation()
      updateSelectedOption(listboxState.items[listboxState.currentIndex])
      closeListBox()
      nextTick(() => focusButton())
      break
    case 'Home':
      event.preventDefault()
      focusAvailableElement(optionsRef.value, (i) => i + 1, 0)
      break
    case 'End':
      event.preventDefault()
      {
        const last = listboxState.items.length - 1
        focusAvailableElement(optionsRef.value, (i) => i - 1, last)
      }

      break
    case 'Escape':
      event.preventDefault()
      closeListBox()
      nextTick(() => focusButton())
      break
    case 'Tab':
      event.preventDefault()
      event.stopPropagation()
      break
    default:
  }
}
const optionsClasses = computed(() => {
  const optionsWrapper: TwClassInterface = {
    position: 'absolute',
    zIndex: 'z-50',
    maxHeight: 'max-h-96',
    minWidth: 'min-w-full',
    overflow: 'overflow-y-auto',
    backgroundColor: 'bg-secondary-50 dark:bg-secondary-900',
    borderWidth: 'border',
    borderColor: 'border-secondary-200 dark:border-secondary-700',
    borderRadius: {
      'rounded-md': props.rounded === true,
      'rounded-2xl': props.rounded === 'full'
    },
    padding: {
      'p-1.5': props.size === 'xs' || props.size === 'sm',
      'p-2': props.size === 'md',
      'p-2.5': props.size === 'lg' || props.size === 'xl'
    },
    boxShadow: 'shadow-lg',
    bottom: properPosition.value == 'top' ? 'bottom-full' : '',
    top: properPosition.value == 'bottom' ? 'top-full' : '',
    margin: properPosition.value == 'bottom' ? 'mt-2' : 'mb-2',
    space: props.size === 'xs' || props.size === 'sm' ? 'space-y-1.5' : 'space-y-2'
  }
  return Object.values({ ...optionsWrapper })
})

const selectWrapperClasses = computed(() => {
  const classes: TwClassInterface = {
    position: 'relative',
    width: props.block ? 'w-full' : ''
    // pointerEvents:
    //   attrs?.disabled !== undefined && attrs.disabled === true
    //     ? "pointer-events-none"
    //     : "",
    // cursor: "cursor-wait",
    // cursor:
    //   attrs?.disabled !== undefined && attrs.disabled === true
    //     ? "cursor-not-allowed"
    //     : "cursor-pointer",
  }
  return Object.values({ ...classes })
})

const inputProps = computed(() => ({
  rounded: props.rounded,
  block: props.block,
  state: props.state,
  stateIcon: props.stateIcon,
  placeholder: props.placeholder,
  size: props.size,
  description: props.description,
  ...attrs
}))

const optionProps = (option: string | object) => {
  const commonProps = {
    size: props.size,
    rounded: props.rounded
  }
  return typeof option === 'string'
    ? { text: option, ...commonProps }
    : { ...option, ...commonProps }
}

const setInputValue = computed(() =>
  typeof selectedOption.value === 'string' ? selectedOption.value : selectedOption.value?.text
)
function arrowIconSize(size: string) {
  return size === 'xs' ? '12' : size === 'sm' ? '16' : size === 'xl' ? '24' : '20'
}
</script>
<template>
  <div
    role="combobox"
    ref="selectWrapperRef"
    aria-haspopup="listbox"
    :class="selectWrapperClasses"
    :aria-expanded="optionsActive"
    :aria-controls="optionsId"
    tabindex="-1"
    @click="toggleOptions"
  >
    <LuiInput
      ref="selectRef"
      :id="selectId"
      v-bind="inputProps"
      :value="setInputValue"
      readonly
      @keydown="buttonKeydown($event)"
    >
      <template v-if="$slots.prepend" #prepend>
        <slot name="prepend" />
      </template>
      <template #append>
        <svg
          viewBox="0 0 12 12"
          :width="arrowIconSize(size)"
          :height="arrowIconSize(size)"
          fill="currentColor"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.99999 6.58599L8.47499 4.11099L9.18199 4.81799L5.99999 7.99999L2.81799 4.81799L3.52499 4.11099L5.99999 6.58599Z"
            fill="currentColor"
          />
        </svg>
      </template>
    </LuiInput>
    <ul
      v-show="optionsActive"
      ref="optionsRef"
      :id="optionsId"
      aria-orientation="vertical"
      :aria-labelledby="selectId"
      role="listbox"
      tabindex="0"
      :class="optionsClasses"
      :aria-activedescendant="listboxState.currentId"
      @keydown="optionsKeydown($event)"
    >
      <LuiOption v-if="placeholder !== ''" disabled :text="placeholder" />
      <template v-if="options.length > 0">
        <LuiOption v-for="(option, index) in options" :key="index" v-bind="optionProps(option)">
        </LuiOption>
      </template>
      <slot v-if="$slots.default" />
    </ul>
  </div>
</template>
