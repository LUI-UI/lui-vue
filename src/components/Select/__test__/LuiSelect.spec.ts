import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import type { DOMWrapper, VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { ref } from 'vue'
import LuiSelect from '../LuiSelect.vue'
import LuiOption from '../../Option/LuiOption.vue'

describe('Default LuiSelect', () => {
  let wrapper: VueWrapper<any>

  beforeEach(() => {
    wrapper = mount(LuiSelect as any)
  })
  afterEach(() => {
    wrapper.unmount()
  })
  it('renders properly', () => {
    const expectedDefProps = {
      rounded: false,
      block: false,
      menuClasses: '',
      menuPosition: 'bottomLeft',
      state: null,
      teleport: false,
      stateIcon: false,
      placeholder: '',
      size: 'md',
      description: null,
      options: [],
      searchable: false,
      showAppend: true,
      modelValue: undefined,
    }
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.html).toMatchSnapshot()
    expect(wrapper.findComponent(LuiSelect).props()).toEqual(expectedDefProps)
  })

  it('displays the placeholder properly', async () => {
    const placeholderExample = 'Names'
    await wrapper.setProps({ placeholder: placeholderExample })

    const inputEl = wrapper.find(
      '[data-testid="lui-select-input"]',
    ) as DOMWrapper<HTMLInputElement>
    expect(inputEl.element.placeholder).toBe(placeholderExample)
  })

  it('renders properly with array options', async () => {
    const exampleOptions = ['Mary', 'Jane']
    await wrapper.setProps({ options: exampleOptions })
    const optionsContainer = wrapper.find(
      '[data-testid="lui-select-options-container"]',
    )

    // test if options are rendered for every options prop
    exampleOptions.forEach((option) => {
      expect(optionsContainer.text()).toContain(option)
    })
  })
  it('renders properly with object array options', async () => {
    const objectArrayOptions = [
      { text: 'Mary', value: 'mary', disabled: true },
      { text: 'Jane', value: 'jane' },
    ]
    await wrapper.setProps({ options: objectArrayOptions })
    const optionsContainer = wrapper.find(
      '[data-testid="lui-select-options-container"]',
    )

    // test if options are rendered for every options prop
    objectArrayOptions.forEach((option) => {
      expect(optionsContainer.text()).toContain(option.text)
    })
  })

  it('disables when disabled prop given', async () => {
    // it is not working with setProps.
    const wrapper = mount(LuiSelect as any, {
      props: {
        disabled: true,
      },
    })
    const inputEl = wrapper.find('[data-testid="lui-select-input"]')
    expect(inputEl.attributes('disabled')).toBeDefined()
  })
  it('works properly with v-model', async () => {
    const selectedName = ref([])
    await wrapper.setProps({
      'options': ['Mary', 'Jane'],
      'onUpdate:modelValue': (e: any) => (selectedName.value = e),
    })

    const options = wrapper.findAll('[data-testid="lui-select-option"]')
    options[0].trigger('click')
    expect(selectedName.value).toEqual('Mary')
  })
})

describe('LuiSelect Search Feature', () => {
  let wrapper: VueWrapper<any>
  let inputEl: DOMWrapper<HTMLInputElement>

  const options = ['Mary', 'Jane']
  const searchKey = 'Mar'
  const falseKey = 'asdsad'

  const testData = {
    options,
    searchKey,
    falseKey,
  }

  beforeEach(() => {
    wrapper = mount(LuiSelect, {
      props: {
        options: testData.options,
        searchable: true,
      },
    })
    inputEl = wrapper.find(
      '[data-testid="lui-select-input"]',
    ) as DOMWrapper<HTMLInputElement>
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('should filter options based on search key', async () => {
    // Initial options
    let options = wrapper.findAll('[data-testid="lui-select-option"]')
    expect(wrapper.vm.searchedOptions.length).toBe(options.length)

    // Search by a key
    await inputEl.setValue(testData.searchKey)

    // Updated options after searching
    options = wrapper.findAll('[data-testid="lui-select-option"]')
    expect(options.length).toBe(1)
    expect(options[0].text()).toBe('Mary')
  })

  it('should display all options when search input is empty', async () => {
    // Clear the search input
    await inputEl.setValue('')

    // Options should be back to their original count
    const options = wrapper.findAll('[data-testid="lui-select-option"]')
    expect(options.length).toBe(testData.options.length)
  })

  it('should display a false search message when no query matches the search key', async () => {
    await inputEl.setValue(testData.falseKey)

    // No options should be displayed
    const options = wrapper.findAll('[data-testid="lui-select-option"]')
    expect(options.length).toBe(0)

    // Check if the false search message is displayed
    const searchError = wrapper.find('[data-testid="lui-select-false-search"]')
    expect(searchError.exists()).toBe(true)
  })
})

describe('LuiSelect initial values', () => {
  let wrapper: VueWrapper<any>

  beforeEach(() => {
    wrapper = mount(LuiSelect as any)
  })
  afterEach(() => {
    wrapper.unmount()
  })

  it('selects the first option from the array when selected is not specified', () => {
    // test for options prop
    const optionsWrapper = mount(LuiSelect as any, {
      props: {
        options: ['Mary', 'Jane'],
      },
    })
    expect(optionsWrapper.vm.selectedOption.value).toBe('Mary')
  })
  it('displays the first option when initial value is not given', () => {
    const slotTemplate = {
      components: { LuiSelect, LuiOption },
      template: `
        <lui-select>
        <lui-option text="Mary" value="Mary"></lui-option>
        <lui-option text="Jane"  value="Jane"></lui-option>
        </lui-select>
        `,
    }
    const slotWrappper = mount(slotTemplate)
    const slotLuiSelect = slotWrappper.findComponent(LuiSelect) as any
    expect(slotLuiSelect.vm.selectedOption.value).toBe('Mary')
  })
  it('selecets initial value properly from v-model value', async () => {
    // const optionsArray = ["Mary", "Jane"];
    // test for v-model
    const modelValue = ref('Mary')
    await wrapper.setProps({
      'modelValue': modelValue.value,
      'onUpdate:modelValue': (e: any) => (modelValue.value = e),
      'options': ['Mary', 'Jane'],
    })
    expect(wrapper.vm.selectedOption.value).toBe(modelValue.value)
  })

  it('selects initial value properly when options are object', async () => {
    const wrapper = mount(LuiSelect as any, {
      props: {
        options: [
          { text: 'Mary', value: 'Mary' },
          { text: 'Jane', value: 'Jane', selected: true },
        ],
      },
    })
    expect(wrapper.vm.selectedOption.value).toBe('Jane')
  })
  it('selects initial value properly with selected slot', async () => {
    const selectedSlotTemplate = {
      components: { LuiSelect, LuiOption },
      template: `
          <lui-select>
          <lui-option text="Mary" value="Mary"></lui-option>
          <lui-option text="Jane" selected value="Jane"></lui-option>
          </lui-select>
          `,
    }
    const selectedSlotWrappper = mount(selectedSlotTemplate)
    const selectedSlotLuiSelect = selectedSlotWrappper.findComponent(
      LuiSelect,
    ) as any
    expect(selectedSlotLuiSelect.vm.selectedOption.value).toBe('Jane')
  })
  it('selects placeholder initially', async () => {
    // test for placeholder
    const placeholder = 'my test placeholder'
    const placeholderWrapper = mount(LuiSelect as any, {
      props: {
        options: ['Mary', 'Jane'],
        placeholder,
      },
    })

    expect(placeholderWrapper.vm.selectedOption.value).toBe(placeholder)
  })
})
describe('LuiSelect with value', () => {
  const optionsWValue = [
    { text: 'Turkey', value: 'TR' },
    { text: 'Austria', value: 'AT' },
    { text: 'Azerbaijan', value: 'AZ' },
    { text: 'Bangladesh', value: 'BD' },
  ]
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiSelect)
  })
  it('accepts value', async () => {
    await wrapper.setProps({ options: optionsWValue })
    const options = wrapper.findAll('[data-testid="lui-select-option"]')
    const optionTexts = options.map(option => option.text())
    const expectedTexts = optionsWValue.map(option => option.text)
    expect(optionTexts).toEqual(expectedTexts)
  })
  it('works with the v-model', async () => {
    const modelValue = ref(optionsWValue[2])
    await wrapper.setProps({
      'options': optionsWValue,
      'modelValue': modelValue.value,
      'onUpdate:modelValue': (e: any) => (modelValue.value = e),
    })
    expect(wrapper.vm.selectedOption.text).toBe('Azerbaijan')

    const options = wrapper.findAll('[data-testid="lui-select-option"]')
    options[3].trigger('click')
    expect(modelValue.value).toBe(optionsWValue[3].value)
  })
})
