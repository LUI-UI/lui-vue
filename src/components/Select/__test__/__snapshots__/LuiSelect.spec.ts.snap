// Vitest Snapshot v1

exports[`Default LuiSelect > renders properly 1`] = `[Function]`;
